vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopWallet.kt

```diff
@@ -119,212 +117,137 @@ class OidcSiopWallet(
-                payload = response.params.serialize().encodeToByteArray(), addX5c = false
-            ).map { it.serialize() }.getOrElse {
-                Napier.w("buildJarm error", it)
-                throw OAuth2Exception(Errors.INVALID_REQUEST)
-            }
-        }
+    /**
+     * Starts the authorization response building process from the RP's [params]
+     */
+    suspend fun startAuthorizationResponsePreparation(
+        params: AuthenticationRequestParametersFrom
+    ): KmmResult<AuthorizationResponsePreparationState> = catching {
+        val clientMetadata = runCatching { params.parameters.loadClientMetadata() }.getOrNull()
+        val presentationDefinition = params.parameters.loadPresentationDefinition()
 
-    private fun RelyingPartyMetadata.requestsEncryption() =
-        authorizationEncryptedResponseAlg != null && authorizationEncryptedResponseEncoding != null
+        AuthorizationRequestValidator(remoteResourceRetriever).validateAuthorizationRequest(params)
 
-    private fun authnResponseQuery(
-        request: AuthenticationRequestParametersFrom,
-        response: AuthenticationResponse
-    ): AuthenticationResponseResult.Redirect {
-        val url = request.parameters.redirectUrl?.let { redirectUrl ->
-            URLBuilder(redirectUrl).apply {
-                response.params.encodeToParameters().forEach {
-                    this.parameters.append(it.key, it.value)
-                }
-            }.buildString()
-        } ?: throw OAuth2Exception(Errors.INVALID_REQUEST)
-        return AuthenticationResponseResult.Redirect(url, response.params)
+        AuthorizationResponsePreparationState(
+            clientMetadata = clientMetadata,
+            presentationDefinition = presentationDefinition,
+        )
     }
 
     /**
-     * That's the default for `id_token` and `vp_token`
+     * Finalize the authorization response
+     *
+     * @param request: the RP's [params]
+     * @param preparationState: The preparation state from [startAuthorizationResponsePreparation]
+     * @param inputDescriptorSubmissions: Map from input descriptor ids to [CredentialSubmission]
      */
-    private fun authnResponseFragment(
+    suspend fun finalizeAuthorizationResponse(
         request: AuthenticationRequestParametersFrom,
-        response: AuthenticationResponse
-    ): AuthenticationResponseResult.Redirect {
-        val url = request.parameters.redirectUrl?.let { redirectUrl ->
-            URLBuilder(redirectUrl)
-                .apply { encodedFragment = response.params.encodeToParameters().formUrlEncode() }
-                .buildString()
-        } ?: throw OAuth2Exception(Errors.INVALID_REQUEST)
-        return AuthenticationResponseResult.Redirect(url, response.params)
+        preparationState: AuthorizationResponsePreparationState,
+        inputDescriptorSubmissions: Map<String, CredentialSubmission>? = null,
+    ): KmmResult<AuthenticationResponseResult> = catching {
```

Comment:
can't we  `map` the result?

Reply:
What exactly do you mean by that?

Reply:
fixed in commit e0d9d1f448615405bd526b2e676580e2b90ae884